## Question:
Use the data set attached to develop machine learning model using Random Forest Classifier and logistic Regression to predict who has Risk to the Bank when doing businesses.

1) Develop a Random forest classifier classifier.(4 marks)

2) Develop a logistic classifier. (4 marks)

3) Evaluate the models. (8 marks)

4) Write your conclusion. (2 marks)

5) Share your work using GitHub URL or upload the codes and write up in this platform (2marks)

Total Marks = 20%

In [1]:
# import pandas library and matplot 
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
## use pandas to load csv file with datasets
data = pd.read_csv('german_credit_data_biased_training.csv')

In [3]:
data.head()

CheckingStatus  LoanDuration           CreditHistory LoanPurpose  \
0       0_to_200            31    credits_paid_to_date       other   
1         less_0            18    credits_paid_to_date     car_new   
2         less_0            15  prior_payments_delayed   furniture   
3       0_to_200            28    credits_paid_to_date  retraining   
4    no_checking            28  prior_payments_delayed   education   

   LoanAmount ExistingSavings EmploymentDuration  InstallmentPercent     Sex  \
0        1889      100_to_500             less_1                   3  female   
1         462        less_100             1_to_4                   2  female   
2         250        less_100             1_to_4                   2    male   
3        3693        less_100          greater_7                   3    male   
4        6235     500_to_1000          greater_7                   3    male   

  OthersOnLoan  ...       OwnsProperty Age  InstallmentPlans Housing  \
0         none  ...  savings_insurance  32              none     own   
1         none  ...  savings_insurance  37            stores     own   
2         none  ...        real_estate  28              none     own   
3         none  ...  savings_insurance  32              none     own   
4         none  ...            unknown  57              none     own   

  ExistingCreditsCount      Job Dependents  Telephone ForeignWorker     Risk  
0                    1  skilled          1       none           yes  No Risk  
1                    2  skilled          1       none           yes  No Risk  
2                    2  skilled          1        yes            no  No Risk  
3                    1  skilled          1       none           yes  No Risk  
4                    2  skilled          1       none           yes     Risk  

[5 rows x 21 columns]

In [4]:
## convert all the non numeric values to numeric values

## checkingStatus convert 
data['CheckingStatus'] = data['CheckingStatus'].astype('category')
data['CheckingStatus']= data['CheckingStatus'].cat.codes

## CreditHistory convert
data['CreditHistory'] = data['CreditHistory'].astype('category')
data['CreditHistory']= data['CreditHistory'].cat.codes

## LoanPurpose convert 
data['LoanPurpose'] = data['LoanPurpose'].astype('category')
data['LoanPurpose']= data['LoanPurpose'].cat.codes

## ExistingSavings convert 
data['ExistingSavings'] = data['ExistingSavings'].astype('category')
data['ExistingSavings']= data['ExistingSavings'].cat.codes

## EmploymentDuration convert 
data['EmploymentDuration'] = data['EmploymentDuration'].astype('category')
data['EmploymentDuration']= data['EmploymentDuration'].cat.codes

## Sex convert 
data['Sex'] = data['Sex'].astype('category')
data['Sex']= data['Sex'].cat.codes

## OthersOnLoan convert 
data['OthersOnLoan'] = data['OthersOnLoan'].astype('category')
data['OthersOnLoan']= data['OthersOnLoan'].cat.codes

## OwnsProperty convert 
data['OwnsProperty'] = data['OwnsProperty'].astype('category')
data['OwnsProperty']= data['OwnsProperty'].cat.codes

## InstallmentPlans convert 
data['InstallmentPlans'] = data['InstallmentPlans'].astype('category')
data['InstallmentPlans']= data['InstallmentPlans'].cat.codes

## InstallmentPlans convert 
data['Housing'] = data['Housing'].astype('category')
data['Housing']= data['Housing'].cat.codes

## Job convert 
data['Job'] = data['Job'].astype('category')
data['Job']= data['Job'].cat.codes


## Telephone convert 
data['Telephone'] = data['Telephone'].astype('category')
data['Telephone']= data['Telephone'].cat.codes

## ForeignWorker convert 
data['ForeignWorker'] = data['ForeignWorker'].astype('category')
data['ForeignWorker']= data['ForeignWorker'].cat.codes

## risk convert 
data['Risk'] = data['Risk'].astype('category')
data['Risk']= data['Risk'].cat.codes



In [6]:
## preview the data if its been converted to numeric values 
data.head()

CheckingStatus  LoanDuration  CreditHistory  LoanPurpose  LoanAmount  \
0               0            31              1            6        1889   
1               2            18              1            2         462   
2               2            15              4            5         250   
3               0            28              1            9        3693   
4               3            28              4            4        6235   

   ExistingSavings  EmploymentDuration  InstallmentPercent  Sex  OthersOnLoan  \
0                0                   3                   3    0             2   
1                3                   0                   2    0             2   
2                3                   0                   2    1             2   
3                3                   2                   3    1             2   
4                1                   2                   3    1             2   

   ...  OwnsProperty  Age  InstallmentPlans  Housing  ExistingCreditsCount  \
0  ...             2   32                 1        1                     1   
1  ...             2   37                 2        1                     2   
2  ...             1   28                 1        1                     2   
3  ...             2   32                 1        1                     1   
4  ...             3   57                 1        1                     2   

   Job  Dependents  Telephone  ForeignWorker  Risk  
0    1           1          0              1     0  
1    1           1          0              1     0  
2    1           1          1              0     0  
3    1           1          0              1     0  
4    1           1          0              1     1  

[5 rows x 21 columns]

In [16]:
## Since we are determining/ predicting a person with high risk when doing business the Risk column will 
## be used as the dependent variable then other column will be the independent variables

X = data.iloc[:,: -1]
y = data.iloc[:,-1].values


In [26]:
X.shape



(5000, 20)

In [27]:
y.shape

(5000,)

In [31]:
## split the dataset into train and test data 

from sklearn.model_selection import train_test_split
X_train,X_test, y_train, y_test = train_test_split(X,y,test_size= 0.3, random_state=0)

# RandomForestClassifier

In [33]:
## classify the model 
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(criterion="gini", max_depth = 8, min_samples_split= 10, random_state = 5)

### Fit the data using the classifier

In [34]:
rfc.fit(X_train,y_train)

RandomForestClassifier(max_depth=8, min_samples_split=10, random_state=5)

## Y predict from X test data

In [35]:
y_pred = rfc.predict(X_test)

In [42]:
y_pred

array([1, 1, 0, ..., 0, 0, 0], dtype=int8)

## Get the acurancy of the model 


In [43]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test,y_pred)

0.7993333333333333

## Classification Report 

In [45]:
from sklearn.metrics import classification_report
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.92      0.81      0.86      1151
           1       0.55      0.78      0.64       349

    accuracy                           0.80      1500
   macro avg       0.74      0.79      0.75      1500
weighted avg       0.84      0.80      0.81      1500





# Logistic Regression classifier


### import model and fit the data into the model

In [46]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train,y_train)

/home/abertnamanya/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

## y predict using LogisticRegression

In [48]:
logic_y_pred = model.predict(X_test)

In [49]:
logic_y_pred

array([1, 1, 0, ..., 1, 1, 0], dtype=int8)

In [51]:
# Get the accurancy of the model 

In [50]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,logic_y_pred))

0.782


## Compaire Classification Report the two models 

In [52]:
# RandomForestClassifier  Claffication Report 
from sklearn.metrics import classification_report
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

           0       0.92      0.81      0.86      1151
           1       0.55      0.78      0.64       349

    accuracy                           0.80      1500
   macro avg       0.74      0.79      0.75      1500
weighted avg       0.84      0.80      0.81      1500



In [53]:
# Logictic Regression Classifier  Claffication Report 
from sklearn.metrics import classification_report
print(classification_report(logic_y_pred,y_test))

              precision    recall  f1-score   support

           0       0.88      0.81      0.84      1091
           1       0.58      0.70      0.64       409

    accuracy                           0.78      1500
   macro avg       0.73      0.76      0.74      1500
weighted avg       0.80      0.78      0.79      1500



## Conclusion: 

#### I would conclude that RandomForestClassifier gives more accurancy of the prediction by 0.80 as for the Logictic
#### Regression Classifier its accurancy is lower thus its predicts may be conclusive compaired to us of 
#### RandomForestClassifier model 